### ISSUE TITLE
Some suggestions for method naming

### ISSUE THREAD

**Issue by leacoleaco** [JOML-CI/JOML#289]
Sometimes overloading  function can lead to some misuse, which is easy to forget even if you read the comment.
for example, these two methods：
```
public Matrix4f invert()
public Matrix4f invert(Matrix4fc dest)
```
When I use it in my projects, chaining calls always make me forget that `invert()` method does not return a new value， so I write these code：
```java
Matrix4f a = new Matrix4f().translation(1,2,3);
Matrix4f b = new Matrix4f().translation(4,5,6);
// I don't want to change the values of a or b in this case
Matrix4f c = b.mul(a.invert());
```
 I don't want to change the values of a or b in this case, but actually I changed the value of a and b. 
the right way I should write is:
```java
Matrix4f c = b.mul(a.invert(new Matrix4f()), new Matrix4f());
```
 I sometimes forget to write the following ‘dest’ arguments，then cause the bug which is not easy to find out。

 
another problem is that the different between `mul()` and `mulLocal()`  method. 
I think the 'preMul' is better than 'mulLocal', because I always mistakenly assumed that `mulLocal()` meant to modify 'local' variables then return, and `mul()` means return a new value.

I read the source code of jmonkeyengin, and its definition is `mulLocal()` modifies itself, and `mul()` returns a new value. see:
https://github.com/jMonkeyEngine/jmonkeyengine/blob/master/jme3-core/src/main/java/com/jme3/math/Matrix4f.java#L1137
https://github.com/jMonkeyEngine/jmonkeyengine/blob/master/jme3-core/src/main/java/com/jme3/math/Matrix4f.java#L1259

Based on my project use, I think use `mulLocal(Matrix4fc right)` or `mulSelf(Matrix4fc right)` instead of `mul(Matrix4fc right)` , use `invertLocal()` or `invertSelf()` instead of `invert()` can give me a better coding experience :-)

so，I suggest：
- `mul(Matrix4fc right)` rename to `mulLocal(Matrix4fc right)`
- `mul(Matrix4fc right, Matrix4fc dest)`  keep naming  `mul(Matrix4fc right, Matrix4fc dest)`
- ` mulLocal(Matrix4fc left)` rename to `preMulLocal(Matrix4fc left)`
- `mulLocal(Matrix4fc left, Matrix4fc dest)` rename to `preMul(Matrix4fc left, Matrix4fc dest)`

or,

-  add a method `mulLocal(Matrix4fc right)` to support `mul(Matrix4fc right)`
- `mul(Matrix4fc right, Matrix4fc dest)`  keep naming  `mul(Matrix4fc right, Matrix4fc dest)`
- ` mulLocal(Matrix4fc left)` rename to `preMul(Matrix4fc left)`
- add method `preMulLocal(Matrix4fc left)`  to support ` preMul(Matrix4fc left)` 
- `mulLocal(Matrix4fc left, Matrix4fc dest)` rename to `preMul(Matrix4fc left, Matrix4fc dest)`

Similar operations should be applied to the other corresponding methods.

This may reduce the number of erroneous calls， I am looking forward to your reply
sincerely

**Comment by httpdigest** [comment]
Thank you for your detailed post.

The issue with JOML's methods modifying `this` by default is definitely only a design choice in favor of better performance, by not relying on escape-analysis to potentially eliminate new object instantiations.

And as you've also shown, this might trip up people when they think that matrices and, even more so, vectors and quaternions should be (immutable) valuable objects - as they really should be - and operations on them produce new values instead of modifying the operated-upon instances.

All of this will be addressed with JOML 2.0 that will rely on Valhalla value types and the Panama Vector APIs.
So: immutable value-types by default and SIMD operations.

But right now, I am afraid we are stuck with the current situation and renaming central methods (like Matrix4f.mul) is a too big API breaking change.

**Comment by leacoleaco** [comment]
Thank you for answer  :-)
